compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -1515,6 +1493,48 @@ static LogicalResult setRootConfig(
   return success();
 }
 
+static LogicalResult setPackOpRootConfig(
+    func::FuncOp entryPointFn, tensor::PackOp op,
+    const TargetMLTransformInfo &targetMLTransInfo,
+    DispatchLoweringPassPipeline pipeline =
+        DispatchLoweringPassPipeline::CPUDataTiling) {
+  TileSizesListType tileSizesList;
+  if (auto srcGenericOp = op.getSource().getDefiningOp<linalg::GenericOp>()) {
+    if (failed(setRootConfig(entryPointFn, srcGenericOp,
+                             LinalgOpInfo(srcGenericOp.getOperation()),
+                             targetMLTransInfo))) {
+      return failure();
+    }
+    tileSizesList = getLoweringConfig(srcGenericOp).getTileSizeVals();
+    pipeline = getTranslationInfo(entryPointFn).getPassPipeline().getValue();
+    removeLoweringConfig(srcGenericOp);
+  } else {
+    SmallVector<int64_t> tileSizes = getLinalgExtDefaultWorkgroupTileSizes(
+        cast<TilingInterface>(op.getOperation()), defaultWorkgroupTileSize);
+    tileSizesList = {tileSizes};
+  }
+
+  // The default function aims to returns the number of workload per workgroup,
+  // but it does not know that it is working on packed domain. We need to take
+  // inner tile sizes into account and adjust the distribution tile sizes.
+  for (SmallVectorImpl<int64_t> &tileSizes : tileSizesList) {
+    SmallVector<int64_t> innerTiles = op.getStaticTiles();
+    ArrayRef<int64_t> dimPos = op.getInnerDimsPos();
+    for (auto [pos, size] : llvm::zip_equal(dimPos, innerTiles)) {
+      if (tileSizes[pos] == 0 || ShapedType::isDynamic(size)) continue;
+      tileSizes[pos] = tileSizes[pos] / size;
+      tileSizes[pos] = std::max<int64_t>(tileSizes[pos], 1);
+    }
+  }
+
+  // The config must be set on tensor.pack ops. Otherwise the
+  // IREE::Flow::DispatchWorkgroupCountFromSetEncodingOp will not get lowered in
+  // tile and distribute pass. See Common/TileAndDistributeToWorkgroupsPass.cpp
+  // for more details.
```

Comment:
I agree removing the lowering config and setting it again is kind of wonky.... Having multiple lowering configs is also a kind of footgun (there are somethings that need to be consistent across lowering configs which is hard to enforce, like having the tile sizes match for tiling levels related to tile + fuse). In this case, I am OK with removing the config and setting it on the pack op as this is a one-off thing. I'd actually not even add a helper function, do it just in the pack op and have big bold warnings in comments as to not do that in any other situation.

Reply:
Sorry, I misunderstood about moving it from one to the other. I was trying to say that the lowering config is consistently carried by pack ops w/ and w/o the change.

>  I don't think that the fact that a pack operation is fused with its producer should change the root op and the overall strategy to apply to the producer.

The pack op **is** the root op. The entry point is `setPackOpRootConfig`. This is the function sets config to a root op, which is a pack op.

Perhaps we have different definition about root op. Given a dispatch function, how do you determine which operation is the root op?  

The overall strategy is not changed either. The pipeline is the same. The loop ranges are basically the same. (If the tiling sizes are multiple of 4 or 8, I think the scaling do nothing on the loop ranges for 99% cases.)

> If for any reason, we had to retrieve the lowering config from the pack operation to apply it to other ops, we would have to scale it back to be applicable to any non-pack operations.

We don't have to scale it back to applicable to any non-pack operations. This is how tile and fuse work:

1. We get tiling sizes from lowering config and set the tiling sizes computation function.
2. The tileAndFuse pattern finds the last tilable op (which can be casted into TilingInterface).
3. The pattern gets the tiling sizes from (1) and apply tiling on the last tilable op. At this moment, the loop body contains extract_slice ops and tiled ops.
4. The algorithm then replaces the extract_slice op with tiled producers using the sizes in the extract_slice ops. The sizes in the extract_slice op are scaled back automatically.

> I think it would be simpler to keep the generic op where it is and then have the pack op to get the lowering config and scale it based on its need. Would that work?

Are you suggesting to create a second pass on dispatch configuration; we can scaling the tile sizes in the second pass? 

-----

Sorry that I don't really get the concern.. Is it:

- The pack op should never be a root op if there are other generic ops?

or

- We should not remove the config from generic op and set it back to pack op?

If the former one is your concern, I'd say that this PR is not related to it because it's already happening. I can send out another PR for fixing your concern.

If the latter one is your concern, is it okay to you that we reset the config back to generic op? I.e., we set the config on generic op, remove the attr, and set the scaled tile sizes back to the generic op in this function.